src/boosting/prediction_early_stop.cpp

```diff
@@ -0,0 +1,88 @@
+#include <LightGBM/prediction_early_stop.h>
+
+using namespace LightGBM;
+
+#include <algorithm>
+#include <vector>
+#include <cmath>
+
+namespace
+{
+  PredictionEarlyStopInstance createNone(const PredictionEarlyStopConfig&)
+  {
+      return PredictionEarlyStopInstance{
+        [](const double*, int)
+        {
+          return false;
+        }, 100000000   // arbitrary number high enough so that the function will rarely be called
+      };
+  }
+
+  PredictionEarlyStopInstance createMulticlass(const PredictionEarlyStopConfig& config)
+  {
+    // marginThreshold will be captured by value
+    const double marginThreshold = config.marginThreshold;
+
+    return PredictionEarlyStopInstance{
+      [marginThreshold](const double* pred, int sz)
+      {
+        // copy and sort
+        std::vector<double> votes(static_cast<size_t>(sz));
+        for (int i=0; i < sz; ++i)
+           votes[i] = pred[i];
+        std::partial_sort(votes.begin(), votes.begin() + 2, votes.end(), std::greater<double>());
```

Comment:
Right now we're not verifying that the prediction vector is at least of size 2, @guolinke would it be ok to check for it here and throw an exception if this is not met?

This means that `PredictRawEarlyStop` would throw.